openmc/tallies.py

```diff
@@ -3522,6 +3527,18 @@ def _create_mesh_subelements(self, root_element):
                         root_element.append(f.mesh.to_xml_element())
                         already_written.add(f.mesh)
 
+    def _create_filter_subelements(self, root_element):
+        already_written = dict()
+        for tally in self:
+            for f in tally.filters:
+                if f not in already_written:
+                    root_element.append(f.to_xml_element())
+                    already_written[f] = f.id
+                else:
+                    # Set the IDs of identical filters with different
+                    # user-defined IDs to the same value
+                    f.id = already_written[f]
```

Comment:
As is, it probably doesn't cause problems because no one has been relying on filter IDs. However, it is a pretty weird side-effect that by writing the tallies to XML, some of the IDs change. I think we should dig a little bit into the MGXS side to see if we can just prevent filters from getting copied, or otherwise ensure that they remain consistent across tallies. Do you have an example of where filters ended up being duplicated?

Good point on `__hash__` and `__eq__`. According to the [Python docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__), it is required that "objects which compare equal have the same hash value", which we are definitely violating in a few places. Let's save that for another day though. I'll open an issue.